### ISSUE TITLE
2FA "Authenticator App" not working in On-premise instance

### ISSUE THREAD

**Issue by jovobe** [bitwarden/web#190]
Today I tried to activate the 2FA with the Authenticator App "Authy" in my self hosted bitwarden Instance and got the following error after entering the generated key from the app: "Errors have occurred Invalid token.".

## Steps to reproduce
1. Login to web vault
2. Click: Settings -> Two-step Login -> Authenticator App
3. Enter Master Password
4. Scan QR-Code with Authy
5. Enter Code generated from Authy
6. Error pops up: "Errors have occurred Invalid token."

## Versions
- Web: `1.23.0`
- /api/version: `1.17.1`

**Comment by kspearrin** [comment]
Make sure your server time is correct.

**Comment by jovobe** [comment]
Checked, it is correct: `Local time: Di 2018-03-20 01:55:40 CET`.

**Comment by kspearrin** [comment]
What about the device you are using Authy on? Those are really the only reasons I know why 2FA would fail. Unless you're just scanning the wrong key for some reason.

**Comment by jovobe** [comment]
The device has also the correct time. Authy works on several other services with this device. Is there a log file somewhere in the bitwarden installation which I can check for a more detailed message?

**Comment by kspearrin** [comment]
I don't think there would be any useful log for that type of error, but all logs are in ./bwdata/core/logs

**Comment by jovobe** [comment]
The log file does not contain any information about the invalid token. Maybe the error is a result of the many upgrades my instance has done so far? Is there a possibility to backup the database, do a fresh installation and import the data back? Or another approach: Is there a way to check the time inside a docker container or the software itself?

**Comment by kspearrin** [comment]
The database wouldn't have anything to do with this type of problem so that is unlikely to help you. You can bash into the `bitwarden-api` container and check the system time there if you want.

**Comment by jovobe** [comment]
That's it! The time inside the docker container is wrong.
`date` executed within the container: Tue Mar 20 01:23:06 UTC 2018
`date` executed from the host: Di 20. Mär 02:23:06 CET 2018

**Comment by kspearrin** [comment]
isnt that the right utc time though?

**Comment by jovobe** [comment]
Yeah, the timezone inside the container is wrong (UTC but should be CET). Is there a way to change it in the docker config?

**Comment by kspearrin** [comment]
The timezones are different, but the time is the same so I am not sure if changing that is going to fix anything. Maybe? I am not sure how to adjust docker container time.

**Comment by jovobe** [comment]
Even changing the timezone in all running containers does not solve the issue. Is there a service in the bitwarden-api container which has to be restarted to load such changes?

**Comment by loozhengyuan** [comment]
I'm encountering the same issue as well. Ever since I had to update Bitwarden, I faced login issues immediately.

**Comment by jovobe** [comment]
> EDIT: By running the update command, it automatically fixes the problem. Even though the time between the container and the host is still different, I was able to get past this issue.

I can't confirm that. On my instance (today updated to 1.25.1 (web) and 1.18.1 (all other containers)) the bug is still there.

I tried fixing it by changing the container timezone to the host timezone before the api service is started (see bugfix here: https://github.com/jovobe/core/commit/edfd0193017d8e2732e7a55b1f4084a4e475fdf5) but even this change does not fix the problem.

**Comment by elliotroy** [comment]
Have you tried setting an env value when creating the contrainer? For example: 
`docker run -e TZ=Europe/Amsterdam debian:jessie date`

**Comment by jovobe** [comment]
> Have you tried setting an env value when creating the contrainer? For example:
`docker run -e TZ=Europe/Amsterdam debian:jessie date`

The timezone is correct (tested by inserting `TZ=Europe/Amsterdam` into `bwdata/env/global.override.env`) but the issue still exists.

**Comment by yassineim** [comment]
Hey you all, you should know the timezone doesn't matter at all. UTC time does. It kind of is the "root" or "origin" time.
